verilog/analysis/checkers/forbid_implicit_declarations_rule_test.cc

```diff
@@ -0,0 +1,153 @@
+
+      // around scope
+      {"module m;assign ", {kToken, "a1"}, " = 1'b1;\nbegin\nend\nendmodule"},
+      {"module m;begin\nend\nassign ", {kToken, "a1"}, " = 1'b1;endmodule"},
+
+      // declaration and assignement separated by begin-end block
+      {"module m;\nwire a1;\nbegin\nend\nassign a1 = 1'b1;\nendmodule"},
+
+      // out-of-scope
+      {"module m;\nbegin wire a1;\nend\nassign ", {kToken, "a1"}, " = 1'b1;\nendmodule"},
+      {"module m;\nbegin wire a1;\nassign a1 = 1'b0;\nend\nassign ", {kToken, "a1"}, " = 1'b1;\nendmodule"},
+      {"module m;\nwire a1;begin assign ", {kToken, "a1"}, " = 1'b0;\nend\nassign a1 = 1'b1;\nendmodule"},
+
+      // multi-level begin-end blocks
+      {"module m;\n"
+       "  wire x1;\n"
+       "  begin\n"
+       "    wire x2;\n"
+       "    begin\n"
+       "      wire x3;\n"
+       "      begin\n"
+       "        wire x4;\n"
+       "        assign x4 = 1'b0;\n"
+       "        assign ", {kToken, "x3"}, " = 1'b0;\n"
+       "        assign ", {kToken, "x2"}, " = 1'b0;\n"
+       "        assign ", {kToken, "x1"}, " = 1'b0;\n"
+       "      end\n"
+       "      assign ", {kToken, "x4"}, " = 1'b0;\n"
+       "      assign x3 = 1'b1;\n"
+       "      assign ", {kToken, "x2"}, " = 1'b0;\n"
+       "      assign ", {kToken, "x1"}, " = 1'b0;\n"
+       "    end\n"
+       "    assign ", {kToken, "x4"}, " = 1'b0;\n"
+       "    assign ", {kToken, "x3"}, " = 1'b0;\n"
+       "    assign x2 = 1'b0;\n"
+       "    assign ", {kToken, "x1"}, " = 1'b0;\n"
+       "  end\n"
+       "  assign ", {kToken, "x4"}, " = 1'b0;\n"
+       "  assign ", {kToken, "x3"}, " = 1'b0;\n"
+       "  assign ", {kToken, "x2"}, " = 1'b0;\n"
+       "  assign x1 = 1'b1;\n"
+       "endmodule"},
+
+       // generate block, TODO: multi-level
+       {"module m;\ngenerate\nendgenerate\nendmodule"},
+       {"module m;\n"
+        "  wire a1;\n"
+        "  assign a1 = 1'b1;\n"
+        "  generate\n"
+        "  endgenerate\n"
+        "  assign a1 = 1'b0;\n"
+        "endmodule"},
+       {"module m;\n"
+        "  generate\n"
+        "    wire a1;\n"
+        "    assign a1 = 1'b1;\n"
+        "  endgenerate\n"
+        "endmodule"},
+       {"module m;\n"
+        "  generate\n"
+        "    assign ", {kToken, "a1"}, " = 1'b1;\n"
+        "  endgenerate\n"
+        "endmodule"},
+       {"module m;\n"
+        "  generate\n"
+        "    wire a1;\n"
+        "    assign a1 = 1'b1;\n"
+        "  endgenerate\n"
+        "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+        "endmodule"},
+       {"module m;\n"
+        "  wire a1;\n"
+        "  assign a1 = 1'b1;\n"
+        "  generate\n"
+        "    assign ", {kToken, "a1"}, " = 1'b1;\n"
+        "  endgenerate\n"
+        "  assign a1 = 1'b0;\n"
+        "endmodule"},
+
+      // TODO: module scope
+
+      // TODO: nets declared inside terminal/port connection list
+  };
```

Comment:
Even more test case ideas based on this passage in LRM 6.10:

_The implicit net declaration shall belong to the scope in which the net reference appears. For example, if the implicit net is declared by a reference in a generate block, then the net is implicitly declared only in that generate block. Subsequent references to the net from outside the generate block or in another generate block within the same module either would be illegal or would create another implicit declaration of a different net (depending on whether the reference meets the preceding criteria). See Clause 27 for information about generate blocks._

Reply:
This lint rule is very close to detecting that all variables are declared in the scope they are used. One about implicit net declaration which yields valid, yet undesirable, Verilog, and the other one is about "no such variable", which is always invalid Verilog.

I would add a test case to verify the scope is proper

```
module m;
    wire a1;
    generate
        wire a2;
        assign a1 = 1'b1; // OK, a1 is in scope
    endgenerate
   assign a2 =1'b1; // Implicit net declaration
endmodule
```

Note that my simulator rejected @ldalek-antmicro's proposed test case because begin/end block is not allowed in a module. I'm trying to find the proper LRM section, but I believe block statement is allowed in processes only.

Reply:
@corco You are correct that bare begin/end blocks were removed from SV grammar (but were in Verilog-2001).  We even have a lint rule for it: https://github.com/google/verible/blob/master/verilog/analysis/checkers/v2001_generate_begin_rule.h

I found that bare begin/end generate blocks were so common in practice, that I supported it in the yacc grammar of verilog.y (and more easily accept legacy verilog), only to provide that lint rule, intended for SV code.  (I don't have dialect selection of rules implemented yet, though.)

Reply:
Okay. I added test case which @corco suggested. Also I replaced `begin-end` blocks with submodules. I got used to them too much and forgot to replace with modules.